File: frontend/packages/contracts/src/strategyStats.ts
import { z } from "zod";
import { IdSchema, PaiseSchema, UtcDateTimeSchema } from "./common";

export const BestNetPnlSchema = z.strictObject({
  runId: IdSchema,
  netPnlPaise: PaiseSchema,
});
export type BestNetPnl = z.infer<typeof BestNetPnlSchema>;

/**
 * Backtest summary per strategy (D26), computed by the backend; screens never aggregate runs.
 * The result fields are null when the strategy has no completed run.
 */
export const StrategyStatsSchema = z
  .strictObject({
    strategyId: IdSchema,
    runsTotal: z.number().int().min(0),
    runsCompleted: z.number().int().min(0),
    runsFailed: z.number().int().min(0),
    /** Queued or running. */
    runsInProgress: z.number().int().min(0),
    lastRunAt: UtcDateTimeSchema.nullable(),
    bestReturnPercent: z.number().nullable(),
    worstReturnPercent: z.number().nullable(),
    winRateMinPercent: z.number().min(0).max(100).nullable(),
    winRateMaxPercent: z.number().min(0).max(100).nullable(),
    worstDrawdownPercent: z.number().lte(0).nullable(),
    bestNetPnl: BestNetPnlSchema.nullable(),
  })
  .refine((s) => s.runsTotal === s.runsCompleted + s.runsFailed + s.runsInProgress, {
    message: "runsTotal must equal completed + failed + in progress",
    path: ["runsTotal"],
  })
  .refine((s) => (s.runsTotal === 0) === (s.lastRunAt === null), {
    message: "lastRunAt is set exactly when there is a run",
    path: ["lastRunAt"],
  })
  .refine(
    (s) => {
      const values = [
        s.bestReturnPercent,
        s.worstReturnPercent,
        s.winRateMinPercent,
        s.winRateMaxPercent,
        s.worstDrawdownPercent,
        s.bestNetPnl,
      ];
      return s.runsCompleted > 0
        ? values.every((v) => v !== null)
        : values.every((v) => v === null);
    },
    { message: "result stats are set exactly when a run completed", path: ["runsCompleted"] },
  )
  .refine(
    (s) =>
      s.bestReturnPercent === null ||
      s.worstReturnPercent === null ||
      s.worstReturnPercent <= s.bestReturnPercent,
    {
      message: "worstReturnPercent must be at most bestReturnPercent",
      path: ["worstReturnPercent"],
    },
  )
  .refine(
    (s) =>
      s.winRateMinPercent === null ||
      s.winRateMaxPercent === null ||
      s.winRateMinPercent <= s.winRateMaxPercent,
    { message: "winRateMinPercent must be at most winRateMaxPercent", path: ["winRateMinPercent"] },
  );
export type StrategyStats = z.infer<typeof StrategyStatsSchema>;

File: backend/libs/nova_contracts/src/nova_contracts/strategy_stats.py
"""StrategyStats (D26): mirrors `frontend/packages/contracts/src/strategyStats.ts`."""

from typing import Annotated, Self

from pydantic import Field, model_validator

from nova_contracts.common import Contract, Id, Paise, UtcDateTime

Count = Annotated[int, Field(ge=0)]
Percent = Annotated[float, Field(ge=0, le=100)]


class BestNetPnl(Contract):
    run_id: Id
    net_pnl_paise: Paise


class StrategyStats(Contract):
    strategy_id: Id
    runs_total: Count
    runs_completed: Count
    runs_failed: Count
    runs_in_progress: Count
    last_run_at: UtcDateTime | None
    best_return_percent: float | None
    worst_return_percent: float | None
    win_rate_min_percent: Percent | None
    win_rate_max_percent: Percent | None
    worst_drawdown_percent: Annotated[float, Field(le=0)] | None
    best_net_pnl: BestNetPnl | None

    @model_validator(mode="after")
    def _rules(self) -> Self:
        if self.runs_total != self.runs_completed + self.runs_failed + self.runs_in_progress:
            raise ValueError("runsTotal must equal completed + failed + in progress")
        if (self.runs_total == 0) != (self.last_run_at is None):
            raise ValueError("lastRunAt is set exactly when there is a run")
        results = (
            self.best_return_percent,
            self.worst_return_percent,
            self.win_rate_min_percent,
            self.win_rate_max_percent,
            self.worst_drawdown_percent,
            self.best_net_pnl,
        )
        expected_set = self.runs_completed > 0
        if any((value is not None) != expected_set for value in results):
            raise ValueError("result stats are set exactly when a run completed")
        best, worst = self.best_return_percent, self.worst_return_percent
        if best is not None and worst is not None and worst > best:
            raise ValueError("worstReturnPercent must be at most bestReturnPercent")
        low, high = self.win_rate_min_percent, self.win_rate_max_percent
        if low is not None and high is not None and low > high:
            raise ValueError("winRateMinPercent must be at most winRateMaxPercent")
        return self
